pkg/monitor/azure/nsg/rulechecker.go

```diff
@@ -0,0 +1,180 @@
+}
+
+func (c ruleChecker) newProperties(addresses []string) properties {
+	return properties{
+		isMaster:         c.isMaster(addresses),
+		isWorker:         c.isWorker(addresses),
+		isAny:            isAny(addresses),
+		isVirtualNetwork: isVirtualNetwork(addresses),
+	}
+}
+
+func (c ruleChecker) isInvalidDenyRule() bool {
+	source := c.newProperties(c.sourceStrings)
+	destination := c.newProperties(c.destinationStrings)
+
+	switch {
+	case source.isNothing() && destination.isNothing():
+		return false
+	case source.isMaster && destination.isMaster:
+		return true
+	case source.isMaster && destination.isWorker:
+		return true
+	case source.isWorker && destination.isMaster:
+		return true
+	case source.isMaster && destination.isAny:
+		return true
+	case source.isAny && destination.isMaster:
+		return true
+	case source.isWorker && destination.isAny:
+		return true
+	case source.isAny && destination.isWorker:
+		return true
+	case source.isAny && destination.isAny:
+		return true
+	case source.isMaster && destination.isVirtualNetwork:
+		return true
+	case source.isVirtualNetwork && destination.isMaster:
+		return true
+	case source.isWorker && destination.isVirtualNetwork:
+		return true
+	case source.isVirtualNetwork && destination.isWorker:
+		return true
+	case source.isVirtualNetwork && destination.isAny:
+		return true
+	case source.isAny && destination.isVirtualNetwork:
+		return true
+	case source.isVirtualNetwork && destination.isVirtualNetwork:
+		return true
+	}
+	return false
+}
+
+func (c ruleChecker) isMaster(addresses []string) bool {
+	prefixes := toPrefixes(c.log, addresses)
+	for _, p := range prefixes {
+		if c.master.Overlaps(p) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c ruleChecker) isWorker(addresses []string) bool {
+	prefixes := toPrefixes(c.log, addresses)
+	for _, wp := range c.workers {
+		for _, p := range prefixes {
+			if wp.Overlaps(p) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isAny(addresses []string) bool {
+	for _, address := range addresses {
+		if address == wildcard {
+			return true
+		}
+	}
+	return false
+}
+
+func isVirtualNetwork(addresses []string) bool {
+	for _, address := range addresses {
+		if address == virtualNetwork {
+			return true
+		}
+	}
+	return false
+}
+
+// toPrefix converts a string value of an IP address or a CIDR range into a prefix
+func toPrefix(value string) (netip.Prefix, error) {
+	if addr, err := netip.ParseAddr(value); err == nil {
+		return netip.PrefixFrom(addr, 32), nil
+	}
+	if prefix, err := netip.ParsePrefix(value); err == nil {
+		return prefix, nil
+	}
+	return netip.Prefix{}, fmt.Errorf("invalid IP address or CIDR range: %s", value)
+}
+
+// toPrefixes converts a slice of addresses into a slice of Prefixes
+func toPrefixes(log *logrus.Entry, addresses []string) []netip.Prefix {
+	prefixes := make([]netip.Prefix, 0, len(addresses))
+	for _, address := range addresses {
+		if address == "*" || address == "VirtualNetwork" {
```

Comment:
Nit: `address == virtualNetwork`

Reply:
Nice catch!  I somehow overlooked that when refactoring.